File: insertion-sort/cpp/data-manual.cpp
#include <iostream>
#include <chrono>

using namespace std;
using namespace std::chrono;

void swapPointer(int *x_pointer, int *y_pointer) {
  int temporary_pointer = *x_pointer;
  *x_pointer = *y_pointer;
  *y_pointer = temporary_pointer;
}

void insertionSort(int array[], int array_length) {
  for(int i = 1; i < array_length; i++) {
    int key = array[i];
    int j = i - 1;
    while( j >= 0 && array[j] > key ) {
      array[j + 1] = array[j];
      j = j - 1;
    }
    array[j + 1] = key;
  }
}

void getSort(int array[], int array_length) {
  int i;
  for(i = 0; i < array_length; i++) {
    cout << array[i] << " ";
  }
}

string calculateMicroSeconds(int duration) {
  int hours = (duration / 60 / 60);
  int minutes = (duration / 60) - hours * 60;
  int seconds = duration - hours * 60 * 60 - minutes * 60;
  string text = to_string(hours) + " jam " + to_string(minutes) + " menit " + to_string(seconds) + " detik.";
  return text;
}

int main() {
  int array[] = {
    2, 10, 50, 11, 1, 6, 4, 20, 17, 14
  };
  int array_length = sizeof(array)/sizeof(array[0]);

  cout << "Data Sebelum di Sorting: \n";
  getSort(array, array_length);

  cout << "\n\n";

  cout << "Data Setelah di Sorting: \n";

  auto start = high_resolution_clock::now();
  insertionSort(array, array_length);
  auto stop = high_resolution_clock::now();
  auto duration = duration_cast<microseconds>(stop - start);

  getSort(array, array_length);

  //cout << "\n\nTotal Waktu Sorting: " << duration.count() << " microseconds.";
  cout << "\n\nTotal Waktu Sorting: " << calculateMicroSeconds(duration.count());
  
  return 0;
}

File: bubble-sort/php/data-manual.php
<?php

function startBubbleSort(&$data) {
  
  $n = sizeof($data);

  for($i = 0; $i < $n; $i++) {
    for($j = 0; $j < $n - $i - 1; $j++) {
      if($data[$j] > $data[$j + 1]) {
        $t = $data[$j];
        $data[$j] = $data[$j + 1];
        $data[$j + 1] = $t;
      }
    }
  }

}

function calculateMicroSeconds($duration) {
  $hours = (int)($duration/60/60);
  $minutes = (int)($duration/60)-$hours*60;
  $seconds = (int)$duration-$hours*60*60-$minutes*60;
  $str = "$hours jam $minutes menit $seconds detik.";
  return $str;
}

$data = [
  1, 29, 12, 49, 96, 21, 17, 90, 34, 54,
  3, 10, 21, 100, 20, 58, 69, 13, 88, 79
];

$len = sizeof($data);

echo "Data Sebelum di Sorting: \n";
for ( $i = 0; $i < $len; $i++ ) {
  #echo "[" . $i + 1 . "]" . $data[$i] . " ";
  echo $data[$i] . " ";
}


echo "\n\n";

echo "Data Setelah di Sorting: \n";

$start = microtime(true);
startBubbleSort($data);
$stop = microtime(true);
$duration = ($stop - $start);

for ( $i = 0; $i < $len; $i++ ) {
  #echo "[" . $i + 1 . "]" . $data[$i] . " ";
  echo $data[$i] . " ";
}

echo "\n\n";
echo "Total Waktu Sorting: " . calculateMicroSeconds($duration);
